### PR TITLE
Add imresize demo using `opencv` mode

### PR DESCRIPTION
**Pull request by pmixer**
To help others avoid struggling like #1906 by giving concrete examples in case not everyone is scipy expert.

### CODE REVIEW

**Comment by takagi** [comment]
@pmixer Sorry for leaving your PR so long. Do you still have an interest in this? Otherwise, please let me take it over. Thanks!

**Comment by pmixer** [comment]
> @pmixer Sorry for leaving your PR so long. Do you still have an interest in this? Otherwise, please let me take it over. Thanks!

Hi @takagi ! Thx for your reply! It is a quite old PR with just a small modification to add sample code for image resize using cupy+cv2, I'd like to close the PR if it's not useful anymore.

**Comment by takagi** [comment]
I think adding an example code would be nice, then please let me review your code.

**Comment by pmixer** [comment]
> I think adding an example code would be nice, then please let me review your code.

Thx! It's just a piece of sample code added to deprecated official doc: https://github.com/cupy/cupy/pull/1915/files

```
import cupyx.scipy.ndimage
import cupy as cp
import cv2

im = cv2.imread('YOUR_IMAGE_PATH')

M = cp.eye(4)
M[0][0] = M[1][1] = 0.5

smaller_shape = (im.shape[0]/2, im.shape[1]/2, 3)
smaller = cp.zeros(smaller_shape) # preallocate memory for resized image
smaller = cupyx.scipy.ndimage.affine_transform(im, M, output_shape=smaller_shape, output=smaller, mode='opencv')
cv2.imwrite('smaller.jpg', cp.asnumpy(smaller)) # smaller image should be saved to your current path
```

I made the modification mainly because the doc just says we could do sth without showing sample code for corresponding part. No modification to cupy source code actually.

File: docs/source/reference/ndimage.rst [hunk 1]
```diff
@@ -27,3 +27,20 @@ OpenCV mode
...
+
+Example to resize image to half size:
+```
+import cupyx.scipy.ndimage
+import cupy as cp
+import cv2
+
+im = cv2.imread('YOUR_IMAGE_PATH')
+
+M = cp.eye(4)
```
Comment (takagi) [inline docs/source/reference/ndimage.rst:39]: nit: Please use lowercase for variables to follow CuPy's style convention.

File: docs/source/reference/ndimage.rst [hunk 2]
```diff
@@ -27,3 +27,20 @@ OpenCV mode
...
+M[0][0] = M[1][1] = 0.5
+
+smaller_shape = (im.shape[0]/2, im.shape[1]/2, 3)
```
Comment (takagi) [inline docs/source/reference/ndimage.rst:42]: Would you use floordiv to make it a tuple of ints?

```suggestion
smaller_shape = (im.shape[0] // 2, im.shape[1] // 2, 3)
```

File: docs/source/reference/ndimage.rst [hunk 3]
```diff
@@ -27,3 +27,20 @@ OpenCV mode
...
+smaller = cp.zeros(smaller_shape) # preallocate memory for resized image
+smaller = cupyx.scipy.ndimage.affine_transform(im, M, output_shape=smaller_shape, output=smaller, mode='opencv')
```
Comment (takagi) [inline docs/source/reference/ndimage.rst:44]: I suppose we don't need to assign the returned value to `smaller` as it is passed to `output` parameter.

```suggestion
cupyx.scipy.ndimage.affine_transform(im, M, output_shape=smaller_shape, output=smaller, mode='opencv')
```

**Comment by takagi** [comment]
It seems that something is changed after you posted this PR and an error occurs when I run the sample code. Let me look into it.

```Python
>>> cupyx.scipy.ndimage.affine_transform(im, M, output_shape=smaller_shape, output=smaller, mode='opencv')
/home/ext-mtakagi/cupy/cupyx/scipy/ndimage/interpolation.py:30: UserWarning: In the current feature the default order of affine_transform is 1. It is different from scipy.ndimage and can change in the future.
  'the future.'.format(func_name))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/ext-mtakagi/cupy/cupyx/scipy/ndimage/interpolation.py", line 212, in affine_transform
    integer_output=integer_output)
  File "cupy/util.pyx", line 103, in cupy.util.memoize.decorator.ret
  File "/home/ext-mtakagi/cupy/cupyx/scipy/ndimage/_interp_kernels.py", line 405, in _get_affine_kernel
    integer_output=integer_output,
  File "/home/ext-mtakagi/cupy/cupyx/scipy/ndimage/_interp_kernels.py", line 269, in _generate_interp_custom
    mode, ixvar, 'xsize_{}'.format(j)))
  File "/home/ext-mtakagi/cupy/cupyx/scipy/ndimage/filters.py", line 720, in _generate_boundary_condition_ops
    return ops
UnboundLocalError: local variable 'ops' referenced before assignment
```

**Comment by pmixer** [comment]
THANK YOU Masayuki Takagi! I’ll recheck the sample code and update the PR later :) Thank you for actively tracking the issue! Pls wait a while and I’ll get it updated in recent days.

Regards，
Zan Huang
________________________________
发件人: Masayuki Takagi <notifications@github.com>
发送时间: Monday, June 29, 2020 12:40:31 AM
收件人: cupy/cupy <cupy@noreply.github.com>
抄送: pmixer <dreaming_hz@hotmail.com>; Mention <mention@noreply.github.com>
主题: Re: [cupy/cupy] Add imresize demo using `opencv` mode (#1915)


It seems that something is changed after you posted this PR and an error occurs when I run the sample code. Let me look into it.

>>> cupyx.scipy.ndimage.affine_transform(im, M, output_shape=smaller_shape, output=smaller, mode='opencv')
/home/ext-mtakagi/cupy/cupyx/scipy/ndimage/interpolation.py:30: UserWarning: In the current feature the default order of affine_transform is 1. It is different from scipy.ndimage and can change in the future.
  'the future.'.format(func_name))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/ext-mtakagi/cupy/cupyx/scipy/ndimage/interpolation.py", line 212, in affine_transform
    integer_output=integer_output)
  File "cupy/util.pyx", line 103, in cupy.util.memoize.decorator.ret
  File "/home/ext-mtakagi/cupy/cupyx/scipy/ndimage/_interp_kernels.py", line 405, in _get_affine_kernel
    integer_output=integer_output,
  File "/home/ext-mtakagi/cupy/cupyx/scipy/ndimage/_interp_kernels.py", line 269, in _generate_interp_custom
    mode, ixvar, 'xsize_{}'.format(j)))
  File "/home/ext-mtakagi/cupy/cupyx/scipy/ndimage/filters.py", line 720, in _generate_boundary_condition_ops
    return ops
UnboundLocalError: local variable 'ops' referenced before assignment

―

**Comment by pmixer** [comment]
Hi @takagi thx for following this thread, I found the problem was caused by my deleting the forked repo.

The sample code has been modified based on your suggestions:

```
import cupyx.scipy.ndimage
import cupy as cp
import cv2

im = cv2.imread('TODO') # pls fill in your image path

trans_mat = cp.eye(4)
trans_mat[0][0] = trans_mat[1][1] = 0.5

smaller_shape = (im.shape[0] // 2, im.shape[1] // 2, 3)
smaller = cp.zeros(smaller_shape) # preallocate memory for resized image

cupyx.scipy.ndimage.affine_transform(im, trans_mat, output_shape=smaller_shape,
                                     output=smaller, mode='opencv')

cv2.imwrite('smaller.jpg', cp.asnumpy(smaller)) # smaller image saved locally
```

tested and made sure it works, here's the colab file:

https://colab.research.google.com/drive/1IARvlEvTSRKyaNVju5n24d0IvXye_euw?usp=sharing

created a new PR #3559 , this could be closed as deleting the forked repo make it hard to track and merge.

**Comment by takagi** [comment]
Thanks! Let me close this one.

**Pull request closed without merging**